src/net_processing.cpp

```diff
@@ -3004,6 +3021,83 @@ void PeerLogicValidation::ConsiderEviction(CNode *pto, int64_t time_in_seconds)
     }
 }
 
+void PeerLogicValidation::EvictExtraOutboundPeers(int64_t time_in_seconds)
+{
+    // Check whether we have too many outbound peers
+    int extra_peers = connman->GetExtraOutboundCount();
+    if (extra_peers > 0) {
+        // If we have more outbound peers than we target, disconnect one.
+        // Pick the outbound peer that least recently announced
+        // us a new block, with ties broken by choosing the more recent
+        // connection (higher node id)
+        NodeId worst_peer = -1;
+        int64_t oldest_block_announcement = std::numeric_limits<int64_t>::max();
+
+        LOCK(cs_main);
+
+        connman->ForEachNode([&](CNode* pnode) {
```

Comment:
Can we loop through mapNodeState here instead? I'd really like to avoid gluing these layers back together.

Also, I believe that would mean that the ```nBlocksInFlight != 0``` check could be applied here.

Reply:
I just wrote this in my other response, but I don't think the blocks in flight check is a good thing to sort on -- I think there may be edge cases where somehow your slowest peer gets protected, and you would evict someone else.

I'll see what it looks like when I try to loop mapNodeState instead.  I think I'll still have to call `ForNode` inside that loop in order to get to the pnode, so I don't really see the difference in design...?  Can you explain a little more what the design goal is for stuff like this?